### PR TITLE
Add class validator resolver

### PR DESCRIPTION
**Pull request by mzalevski**
This PR adds `class-validator` resolver with basic tests. In addition it fixes the prettier configPath and the copy function.

### CODE REVIEW

**Comment by mzalevski** [comment]
@bluebill1049 @jorisre I am not sure why these checks are failing. I've changed [this](https://github.com/react-hook-form/resolvers/pull/147/commits/82beaf5c7183416b7d748c475ca2d6a0bc9b95ca) to make `yarn build` work locally. Then after the checks failed for the first time, I reverted that change [here](https://github.com/react-hook-form/resolvers/pull/147/commits/37889ddfd4129976f74d8fc2cf2f4e75cba3f120). Sadly the checks are still failing. Any advice?

File: class-validator/src/class-validator.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  schemaOptions = {},
+  resolverOptions = {},
+) => async (values, _, options) => {
+  const user = plainToClass(schema, values);
+  const rawErrors =
+    resolverOptions.mode === 'sync'
+      ? validateSync(user, schemaOptions)
+      : await validate(user, schemaOptions);
+  const errors = toNestError(parseErrors(rawErrors), options.fields);
+  return { values, errors };
```
Comment (jorisre) [inline class-validator/src/class-validator.ts:32]: if there is errors, the resolver should return:
```ts
return { values: {}, errors }
```

if not:
```ts
return { values, errors: {} }
```

**Review by jorisre — COMMENTED** [review]
That's awesome to support another validation library, thanks a lot @snmgian for your contribution ❤️ 

Can you add a small example in the readme too ? thanks !

**Comment by jorisre** [comment]
For the error:
>The 'this' keyword is equivalent to 'undefined' at the top level of an ES module, and has been rewritten

I can't find a way to fix it, we are using microbundle and the rollup context options isn't exposed
https://rollupjs.org/guide/en/#error-this-is-undefined

File: class-validator/package.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  "private": true,
+  "description": "React Hook Form validation resolver: class-validator",
+  "main": "dist/class-validator.js",
+  "module": "dist/class-validator.module.js",
+  "umd:main": "dist/class-validator.umd.js",
+  "source": "src/index.ts",
+  "types": "dist/index.d.ts",
+  "license": "MIT",
+  "peerDependencies": {
+    "react-hook-form": ">=6.6.0",
```
[the remark this answers is not in the document]
Comment (mzalevski) [reply class-validator/package.json:14]: Thanks. I've merged #148

File: class-validator/src/class-validator.ts (hunk 1 above)
Comment (mzalevski) [reply class-validator/src/class-validator.ts]: done [here](https://github.com/react-hook-form/resolvers/pull/147/commits/4195b56508407b7cf9764d653043043ed8f66317)

**Comment by bluebill1049** [comment]
wow, this is awesome! another resolver ~

File: class-validator/src/class-validator.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import type { Resolver } from './types';
+import { plainToClass } from 'class-transformer';
+import { validate, validateSync, ValidationError } from 'class-validator';
+import { toNestError } from '@hookform/resolvers';
+import { FieldError } from 'react-hook-form';
+
+const parseErrors = (
+  rawErrors: ValidationError[],
+): Record<string, FieldError> => {
+  const errors = Object.fromEntries(
```
[the remark this answers is not in the document]
Comment (mzalevski) [reply class-validator/src/class-validator.ts:10]: I've added an example [here](https://github.com/react-hook-form/resolvers/pull/147/commits/6de8bf1d7db06a2b1c0c59c2c3ae8987773caeba) :)

[the remark this answers is not in the document]
Comment (mzalevski) [reply class-validator/src/class-validator.ts]: I've implemented es2017 compliant Object.fromEntries [here](https://github.com/react-hook-form/resolvers/pull/147/commits/af9b0099d954b87f4f70bab62ea62ed74c168243)

File: class-validator/src/__tests__/__snapshots__/class-validator.ts.snap [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      },
+    ],
+    "password": "___",
+  },
+}
+`;
+
+exports[`classValidatorResolver should return a single error from classValidatorResolver with \`mode: sync\` when validation fails 1`] = `
+Object {
+  "errors": Object {
```
[the remark this answers is not in the document]
Comment (mzalevski) [reply class-validator/src/__tests__/__snapshots__/class-validator.ts.snap:43]: I've fixed nested schemas validation [here](https://github.com/react-hook-form/resolvers/pull/147/commits/5f747ff36289cc1b06a7c6bf5c793b674ee528ec)

File: class-validator/src/__tests__/__snapshots__/class-validator.ts.snap [hunk 5]
```diff
@@ -0,0 +1,79 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`classValidatorResolver should return a single error from classValidatorResolver when validation fails 1`] = `
+Object {
+  "errors": Object {
```
[the remark this answers is not in the document]
Comment (mzalevski) [reply class-validator/src/__tests__/__snapshots__/class-validator.ts.snap:5]: I've fixed nested schemas validation [here](https://github.com/react-hook-form/resolvers/pull/147/commits/5f747ff36289cc1b06a7c6bf5c793b674ee528ec)

File: class-validator/src/__tests__/__snapshots__/class-validator.ts.snap (hunk 4 above)
[the remark this answers is not in the document]
Comment (jorisre) [reply class-validator/src/__tests__/__snapshots__/class-validator.ts.snap:43]: Thanks ! So we have to use `@ValidateNested({ each: true })` to get nested schema errors ?

[the remark this answers is not in the document]
Comment (mzalevski) [reply class-validator/src/__tests__/__snapshots__/class-validator.ts.snap:43]: Correct. `{ each: true }` only for arrays though. Other thing is that my `parseErrors` fn was flawed and even with this decorator it did not return all errors properly.

File: class-validator/src/class-validator.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mzalevski) [reply class-validator/src/class-validator.ts]: "I'm new to class-validator and I don't know why rawError.constraints could be undefined, do you know why ?" I dont :/

File: class-validator/src/__tests__/__snapshots__/class-validator.ts.snap (hunk 4 above)
[the remark this answers is not in the document]
Comment (jorisre) [reply class-validator/src/__tests__/__snapshots__/class-validator.ts.snap:43]: Good to know, thanks

File: class-validator/src/class-validator.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mzalevski) [reply class-validator/src/class-validator.ts]: @jorisre I wll try to implement `criteriaMode === "all"` later. Do you have and idea how to make the GH actions checks run successfully? It is something related to kebab-case'ed `class-validator`, isn't it?

[the remark this answers is not in the document]
Comment (jorisre) [reply class-validator/src/class-validator.ts]: It's related to the error:
>The 'this' keyword is equivalent to 'undefined' at the top level of an ES module, and has been rewritten

https://github.com/react-hook-form/resolvers/pull/147#issuecomment-815589323

I've to investigate

[the remark this answers is not in the document]
Comment (jorisre) [reply class-validator/src/class-validator.ts]: No worries for the `criteriaMode === "all"`, I can do it if you want ;)

[the remark this answers is not in the document]
Comment (mzalevski) [reply class-validator/src/class-validator.ts]: ok, @jorisre you can do `criteriaMode === "all"`. I figure that I won't be able to help with `this keyword is equicalent to 'undefined'` errors. So is there anything else that I need to fix? :)

[the remark this answers is not in the document]
Comment (jorisre) [reply class-validator/src/class-validator.ts]: I'll review your changes :)

To get your PR merged, we have to find a solution for the error at build time if not it couldn't be published on NPM

[the remark this answers is not in the document]
Comment (mzalevski) [reply class-validator/src/class-validator.ts]: @jorisre I am sure there is a way to do it. Afterall `class-validator` & `class-transformer` are on npm.

**Review by jorisre — DISMISSED** [review]
LGTM, Thanks a lot @mzalevski  !

**Review by jorisre — APPROVED** [review]

**Comment by mzalevski** [comment]
@jorisre I updated the package and for some reason nested validation does not return errors :/

**Comment by jorisre** [comment]
> @jorisre I updated the package and for some reason nested validation does not return errors :/ I am double checking it right now.

I know why, give me some time to fix it. I'll add criteria mode all too

**Comment by mzalevski** [comment]
Thanks :)

**Pull request merged**